gitlab ci: Rework how mirrors are configured

scottwittenburg opened the pull request:
Improve how mirrors are used in gitlab ci, where we have until now thought of them as only a string.  This PR aims to catch gitlab ci up to where the rest of spack is, in terms of mirror configuration.

By configuring ci mirrors ahead of time using the proposed mirror templates, and by taking advantage of the expressiveness that spack now has for mirrors, this PR will allow us to easily switch the protocol/url we use for fetching binary dependencies.

This PR also:

- Adds three mirror templates which, assuming some variable replacement, should account for all the ways we use mirrors in ci.
- Deprecates complicated (and likely somehow wrong) mirror management logic from the ci module.
- Deprecates `temporary storage` and `artifacts buildcache` functionality from gitlab pipeline schema.
- Deprecates `--buildcache-destination` and `--copy-to` options from the `spack ci generate` command.

Still TODO:

- [x] create a spackbot PR to set the `PIPELINE_MIRROR_TEMPLATE` appropriately for "rebuild everything" pipelines (done [here](https://github.com/spack/spackbot/pull/92))
- [x] add backwards compatibility for things the community is actually using
- [x] set up cloudfront for s3://spack-binaries-prs (and fix urls in mirror templates)
- [x] to avoid generating pipelines from stale indices, tell cloudfront not to cache binary indices (done in #40339)
- [x] solve problem where already cached things ignore re-writing origin objects with `CacheContent: no-cache` headers
  * For `develop` mirror,I think we can write a script to issue cache invalidations for the existing index files, followed immediately by updating the index using the `CacheContent: no-cache` header.  Since caching only happens on client requests, we should not have to worry about the PR mirrors as long as we merge #40339 before this.

scottwittenburg commented:
fyi @kwryankrattiger @zackgalbreath  I finally made this PR from the branch I've been using to test fetching from cloudfront in pipelines.

I'm still not sure how we bust the cache managed by cloudfront when we:

- prune specs from mirrors (didn't come up in my testing)
- update buildcache index in between pipelines (didn't come up in my testing where every pipeline was a "rebuild everything")

kwryankrattiger commented:
For the first point, I agree we need to investigate more. I can imagine some edge cases where it would be problematic and pipelines would fail where a spec hash left existence two weeks earlier, and was reintroduced via a reversion or something on a PR that was depending on the old cached binary. I can't think of a case where it would break develop pipelines though, but maybe you have some ideas.

For the second one I am less concerned about the growing mirror case. In the worst case it will schedule redundant jobs and then do nothing when it finds the binary in the cache. It isn't ideal, but it shouldn't break pipelines which I think is the most important. And if we are getting a huge speed bump in fetch times with CF, the cost of extra scheduled jobs may be outweighed. That may be something worth measuring and/or annotating in the database so we can compare how many jobs like that are getting scheduled and how long they are taking to run. It may be that CF causes a significant overhead.

In the case where things are removed from the index, I think that follows the first point.

scottwittenburg commented:
If it's possible to avoid caching at all for certain patterns (`.+index.json$` or similar), I wonder if that would solve both points?

scottwittenburg committed:
gitlab ci: Rework how mirrors are configured

Improve how mirrors are used in gitlab ci, where we have until now thought
of them as only a string.

By configuring ci mirrors ahead of time using the proposed mirror templates,
and by taking advantage of the expressiveness that spack now has for mirrors,
this PR will allow us to easily switch the protocol/url we use for fetching
binary dependencies.

This change also deprecates some gitlab functionality and marks it for
removal in Spack 0.23:

    - arguments to "spack ci generate":
        * --buildcache-destination
        * --copy-to
    - gitlab configuration options:
        * enable-artifacts-buildcache
        * temporary-storage-url-prefix

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/configs/multi-src-mirrors.yaml.in:
```diff
@@ -0,0 +1,16 @@
+mirrors:
+  buildcache-source:
+    fetch: ${PROTECTED_MIRROR_FETCH_DOMAIN}/${PR_TARGET_REF_NAME}/${SPACK_CI_STACK_NAME}
+    push: s3://spack-binaries/${PR_TARGET_REF_NAME}/${SPACK_CI_STACK_NAME}
```
Could this be `${PROTECTED_MIRROR_PUSH_DOMAIN}`? Then we could switch between staging/production from one place.

scottwittenburg replied:
Yes, I've been thinking all domains should be variables, and then we should have all four possibilities: push/fetch x protected/pr

scottwittenburg commented:
[Fixed](https://github.com/spack/spack/compare/91c7aa92929df3b625ef4f7758803633a50f5970..7ca2028d6b52b1e5fbaf83f94d2975702d655f72) one [issue](https://gitlab.staging.spack.io/spack/spack/-/jobs/15289) caught by testing in staging.

scottwittenburg commented:
Used an old PR on `spack-test/spack` to test the proposed changes to spackbot (to make sure we can still "rebuild everything").  It's [here](https://github.com/spack-test/spack/pull/41#issuecomment-1756232418).

The proof that it worked is, e.g., [here](https://gitlab.spack.io/scott/pipeline-experiments/-/jobs/8616290#L49), where you can see that instead of the usual three mirrors you'd expect for a PR pipeline, there's just the one, and it corresponds to the PR-specific version.

scottwittenburg commented:
I did some more testing of the various flavors of pipeline in my side repo, which uncovered a couple of issues I have now fixed.

- PR [pipeline](https://gitlab.spack.io/scott/pipeline-experiments/-/pipelines/520322) rebuilding only what changed (`readline` in `build_systems`)
- PR [pipeline](https://gitlab.spack.io/scott/pipeline-experiments/-/pipelines/520342) rebuilding everything (triggered from [here](https://github.com/spack-test/spack/pull/41#issuecomment-1760530451))
- a `develop` [pipeline](https://gitlab.spack.io/scott/pipeline-experiments/-/pipelines/520362) rebuilding only what changed 
- a "develop snapshot" [pipeline](https://gitlab.spack.io/scott/pipeline-experiments/-/pipelines/520382) (copy-only)
- a release branch [pipeline](https://gitlab.spack.io/scott/pipeline-experiments/-/pipelines/520402) (rebuild everything)
- a release tag [pipeline](https://gitlab.spack.io/scott/pipeline-experiments/-/pipelines/520427) (copy-only)

All those were run with some extra WIP commits to:

- build many fewer specs in each stack
- change all the mirror urls to point away from production mirrors
- comment out the `id_tokens` bits so we could fall back on aws credentials from ci variables

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -98,8 +106,8 @@ default:
       variables:
         SPACK_PIPELINE_TYPE: "spack_copy_only"
-        SPACK_SOURCE_MIRROR: "s3://spack-binaries/SPACK_REPLACE_VERSION/${SPACK_CI_STACK_NAME}"
         SPACK_COPY_BUILDCACHE: "s3://spack-binaries/${CI_COMMIT_REF_NAME}"
```
`SPACK_COPY_BUILDCACHE: "${PROTECTED_MIRROR_PUSH_DOMAIN}/${CI_COMMIT_REF_NAME}"`

kwryankrattiger commented on lib/spack/spack/ci.py:
```diff
@@ -775,17 +776,33 @@ def generate_gitlab_ci_yaml(
+        mirror_urls = [url for url in ci_mirrors.values()]
+        remote_mirror_url = mirror_urls[0]
 
     spack_buildcache_copy = os.environ.get("SPACK_COPY_BUILDCACHE", None)
```
Should this still be guarded by the environment variable?

scottwittenburg replied:
I don't follow, which environment variable should be guarding what?

kwryankrattiger replied:
The existence of `SPACK_COPY_BUILDCACHE` is still the activation requirement for writing the manifest file. The new method uses known named mirrors to determine the copy location. I think that this should be fixed.

scottwittenburg replied:
I think you might be confusing copy-only pipelines (where we now use known named mirrors to know the copy destination), with this env var, which is used to trigger writing the manifest file.

kwryankrattiger replied:
Ah, right, I am merging our other conversion about the manifest file into this PR. We can address this later.

kwryankrattiger commented on lib/spack/spack/ci.py:
```diff
@@ -805,13 +823,15 @@ def generate_gitlab_ci_yaml(
+    # TODO: Remove this block in Spack 0.23
     mirrors_to_check = None
-    if remote_mirror_override:
+    if deprecated_mirror_config and remote_mirror_override:
```
I am not sure if it makes sense to disable this based on the `deprecated_mirror_config` flag since it is still a CLI option.

scottwittenburg replied:
I don't want to modify the mirror state unless the user didn't use the new approach of adding a mirror named `buildcache-destination` and instead used that command line option.  Did I get that logic wrong?

kwryankrattiger replied:
The option `--buildcache-destination` can still be specified without error, even in the new method. I think either this should be an error earlier saying we don't support both, or this specific override method should be allowed even in the new mode.

I didn't comment on the other mirror gymnastics we were doing because those were sort of magic from the outside world perspective.

scottwittenburg replied:
Ok, I'm leaning toward an error saying we don't support both at the same time, unless you object.

kwryankrattiger replied:
That is also my inclination, having both active at the same time feels odd.

kwryankrattiger commented on lib/spack/spack/ci.py:
```diff
@@ -1249,20 +1299,24 @@ def main_script_replacements(cmd):
 
-        if remote_mirror_override:
+        # TODO: Remove this block in Spack 0.23
+        if deprecated_mirror_config and remote_mirror_override:
```
see previous comment on override mirror

scottwittenburg replied:
Same response.  Maybe you can be more clear what your concern is?

kwryankrattiger replied:
Resolves in prior discussion.

kwryankrattiger commented on share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:
```diff
@@ -68,15 +74,17 @@ default:
-    SPACK_BUILDCACHE_DESTINATION: "s3://spack-binaries/${CI_COMMIT_REF_NAME}/${SPACK_CI_STACK_NAME}"
+    PIPELINE_MIRROR_TEMPLATE: "single-src-protected-mirrors.yaml.in"
+    # TODO: We can remove this when we drop the "deprecated" stack
+    PUSH_BUILDCACHE_DEPRECATED: "s3://spack-binaries/${CI_COMMIT_REF_NAME}/${SPACK_CI_STACK_NAME}"
```
I think this needs the `${PROTECTED_MIRROR_PUSH_DOMAIN}` substituted as well.

kwryankrattiger reviewed (commented):
I still need to verify the stacks have removed the mirror section, but having the old mirror listed also shouldn't hurt anything.

kwryankrattiger reviewed (approved):
This looks good, thanks @scottwittenburg !
